src/protocols/secure_channel/CASESession.cpp

```diff
@@ -339,57 +300,54 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session, SecureSessi
 
-    // Start writing TLV
+    // Fill in the random value
+    ReturnErrorOnFailure(DRBG_get_bytes(initiatorRandom, kSigmaParamRandomNumberSize));
+
+    // Construct Sigma1 Msg
+    msg_R1 = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_ERROR_NO_MEMORY);
+
     tlvWriter.Init(std::move(msg_R1));
     ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-    ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kRandom, initiatorRandom, sizeof(initiatorRandom)));
-
-    // Step 5
-    uint16_t n_trusted_roots = mOpCredSet->GetCertCount();
-    // Initiator's session ID
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kSessionID, mConnectionState.GetLocalKeyID(), true));
-    // Step 2/3
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kNumberofTrustedRootIDs, n_trusted_roots, true));
-    for (uint16_t i = 0; i < n_trusted_roots; ++i)
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(1), initiatorRandom, sizeof(initiatorRandom)));
+    // Retrieve Session Identifier
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), mConnectionState.GetLocalKeyID(), true));
+    // Generate a Destination Identifier
     {
-        CertificateKeyId trustedRootId = mOpCredSet->GetTrustedRootId(i);
-        if (!trustedRootId.empty())
-        {
-            ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kTrustedRootID, trustedRootId.data(), trustedRootId.size()));
-        }
+        ChipCertificateData nodeOperationalCertificate;
+        FabricId fabricId;
+        MutableByteSpan destinationIdSpan(destinationIdentifier, sizeof(destinationIdentifier));
+
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(mTrustedRootId), mOpCredSet->GetDevOpCredLen(mTrustedRootId),
+                                            nodeOperationalCertificate));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
+        ReturnErrorOnFailure(GenerateDestinationID(ByteSpan(initiatorRandom, sizeof(initiatorRandom)), &mTrustedRootId,
+                                                   mConnectionState.GetPeerNodeId(), fabricId, destinationIdSpan));
     }
-    ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kInitiatorEphPubKey, mEphemeralKey.Pubkey(),
-                                            static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(3), destinationIdentifier, sizeof(destinationIdentifier)));
 
+    ReturnErrorOnFailure(
+        tlvWriter.PutBytes(TLV::ContextTag(4), mEphemeralKey.Pubkey(), static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
     ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
     ReturnErrorOnFailure(tlvWriter.Finalize(&msg_R1));
 
     ReturnErrorOnFailure(mCommissioningHash.AddData(msg_R1->Start(), msg_R1->DataLength()));
```

Comment:
@balducci-apple We are computing transcript on pre-header payloads here and not including the headers, and at receiver, just on the payload as well. Please provide a link to spec issue where we clarify which is which.

For now, I consider this "to-spec", as written, but the intent is clearly not respected fully since parts of the message (headers) are not under transcript.